calibrate-touchscreen/src/main.rs

```diff
@@ -0,0 +1,429 @@
+                            wl_surface: None,
+                            buffer: None,
+                            xdg_surface: None,
+                    }));
+
+                    state.draw_fullscreen_surfaces(qh);
+                }
+            }
+
+            _ => {}
+        }
+    }
+}
+
+fn draw_target(tmp: &mut File, (buf_x, buf_y): (u32, u32), (target_x, target_y): (f64, f64)) {
+    let centre_x = (buf_x as f64 * target_x) as i64;
+    let centre_y = (buf_y as f64 * target_y) as i64;
+    let target_size = (max(buf_x, buf_y)/80) as i64;
+    use std::{io::Write};
+    let mut buf = std::io::BufWriter::new(tmp);
+    for y in 0..buf_y as i64 {
+        for x in 0..buf_x as i64 {
+            let distance_squared =(x-centre_x)*(x-centre_x) + (y-centre_y)*(y-centre_y);
+            if distance_squared >= target_size*target_size {
+                let a = 0xFF;
+                let r = 0x3F;
+                let g = 0x3F*y as u32/buf_y;
+                let b = 0x3F*y as u32/buf_y;
+                buf.write_all(&[b as u8, g as u8, r as u8, a as u8]).unwrap();
+            }
+            else {
+                let intensity = 0xFF - 0x40*(distance_squared/(target_size*target_size/4));
+                let a = 0xFF;
+                buf.write_all(&[intensity as u8, intensity as u8, intensity as u8, a as u8]).unwrap();
+            }
+        }
+    }
+    buf.flush().unwrap();
+}
+
+impl State {
+    fn draw_fullscreen_surfaces(&mut self, qh: &QueueHandle<State>) {
+
+        if self.compositor.is_none() || self.wm_base.is_none() || self.wl_shm.is_none() {
+            return;
+        }
+
+        let compositor = self.compositor.as_ref().unwrap();
+        let wm_base = self.wm_base.as_ref().unwrap();
+        let wl_shm = self.wl_shm.as_ref().unwrap();
+
+        for output in self.outputs.iter_mut() {
+
+            if output.1.is_none() {
+                return;
+            }
+
+            let window = output.1.as_mut().unwrap();
+
+            if window.wl_surface.is_none() {
+                let surface = compositor.create_surface(qh, ());
+                window.wl_surface = Some(surface);
+            }
+
+            if window.xdg_surface.is_none() {
+                let wl_surface = window.wl_surface.as_ref().unwrap();
+                let xdg_surface = wm_base.get_xdg_surface(wl_surface, qh, ());
+                let toplevel = xdg_surface.get_toplevel(qh, ());
+
+                // let output = &unsafe { *window.output };
+                toplevel.set_title(output.0.id().to_string());
+
+                toplevel.set_fullscreen(Some(output.0));
+
+                wl_surface.commit();
+
+                window.xdg_surface = Some((xdg_surface, toplevel));
+            }
+
+            let wl_surface = window.wl_surface.as_ref().unwrap();
+
+            let mut file = tempfile::tempfile().unwrap();
+            let width = window.width as u32;
+            let height = window.height as u32;
```

Comment:
I think the idiomatic way to write this is
```suggestion
            let width = u32::try_from(window.width)?;
            let height = u32::try_from(window.height)?;
```
(assuming the function has been refactored to return a `Result<(), Box<dyn Error>>`).